lib/stdlib/doc/src/uri_string_usage.xml

```diff
@@ -0,0 +1,323 @@
+    </p>
+    <pre>
+    6> uri_string:percent_decode("local%23host").
+    "local#host"
+    7> uri_string:percent_decode("/%F6re%26bro%20").
+    <![CDATA[{error,invalid_utf8,<<"/öre&bro ">>}]]>
+    8> uri_string:percent_decode(#{host => "local%23host",path => "/%F6re%26bro%20",
+    scheme => "http"}).
+    <![CDATA[#{host => "local#host",
+      path => {error,invalid_utf8,<<"/öre&bro ">>},
+      scheme => "http"}]]>
+    </pre>
+    <p>The <c>host</c> was successfully decoded but the path contains at least one
+    character with
+    non-UTF-8 encoding. In order to be able to decode this, we have to make assumptions
+    about the encoding used in these triplets. The most obvious choice is
+    <i>latin-1</i>, so we can try
+    to transcode the path to UTF-8 and run the percent-decode operation on
+    the transcoded string:
+    </p>
+    <pre>
+    9> uri_string:transcode("/%F6re%26bro%20", [{in_encoding, latin1}]).
+    "/%C3%B6re%26bro%20"
+    10> uri_string:percent_decode("/%C3%B6re%26bro%20").
+    <![CDATA["/öre&bro "]]>
+    </pre>
+    <p>It is important to emphasize that it is not safe to apply
+    <c>uri_string:percent_decode/1</c> directly on an input URI:
+    </p>
+    <pre>
+    11> uri_string:percent_decode("http://%6C%6Fcal%23host/%C3%B6re%26bro%20").
+    <![CDATA["http://local#host/öre&bro "
+    12> uri_string:parse("http://local#host/öre&bro ").]]>
+    {error,invalid_uri,":"}
+    </pre>
+    <p>This is the reason why OTP depricated the <c>http_uri:decode/1</c> function.
+    It can change a URI or render it invalid. <c>http_uri:encode/1</c> is depricated
+    as well, encoding happens in <c>uri_string:recompose/1</c>.
+    </p>
+  </section>
+
+  <section>
+    <title>Normalization</title>
+    <p>Normalization is the operation of converting the input URI into a <i>canonical</i>
+    form and keeping the reference to the same underlying resource. The most common
+    application of normalization is determining whether two URIs are equivalent
+    without accessing their referenced resources.</p>
+    <p>Normalization has 6 distinct steps. First the input URI is parsed into an
+    intermediate form that can handle Unicode characters. This datatype is the
+    <c>uri_map()</c>, that can hold the components of the URI in map elements of type
+    <c>unicode:charstring()</c>. After having the intermediate form, a sequence of
+    normalization algorithms applied to the individual URI components:</p>
+    <list type="bulleted">
+      <item><p>
+	<em>Case normalization</em>: converts the <c>scheme</c> and <c>host</c> components
+	to lower case as they are not case sensitive.
+      </p></item>
+      <item><p>
+	<em>Percent-encoding normalization</em>: decodes percent-encoded triplets that
+	correspond to characters in the unreserved set.
+      </p></item>
+      <item><p>
+	<em>Scheme-based normalization</em>: applying rules for the schemes http, https,
+	ftp, ssh, sftp and tftp.
```

Comment:
You should probably mention these at the top of the page as well, it's not *just* the URI RFC, some scheme-specific rules from other RFCs also apply.

Reply:
I have to leave this open for now as I'm unable to find the source that I used for https, ftp, ssh, sftp and tftp. It can be that I generalized the rules of [RFC3986 - 6.2.3.  Scheme-Based Normalization](https://tools.ietf.org/html/rfc3986#section-6.2.3) and applied it on these schemes to normalize the port numbers.